src/main/java/io/supertokens/session/Session.java

```diff
@@ -104,19 +110,25 @@ public static SessionInformationHolder createNewSession(Main main, @Nonnull Stri
      *
      */
     public static SessionInformationHolder regenerateToken(Main main, @Nonnull String token,
-            @Nullable JsonObject userDataInJWT) throws StorageQueryException, StorageTransactionLogicException,
-            UnauthorisedException, InvalidKeySpecException, SignatureException, NoSuchAlgorithmException,
-            InvalidKeyException, UnsupportedEncodingException {
+            @Nullable JsonObject userDataInJWT, @Nullable JsonObject grantPayload) throws StorageQueryException,
+            StorageTransactionLogicException, UnauthorisedException, InvalidKeySpecException, SignatureException,
+            NoSuchAlgorithmException, InvalidKeyException, UnsupportedEncodingException {
 
         // We assume the token has already been verified at this point. It may be expired or JWT signing key may have
         // changed for it...
         AccessTokenInfo accessToken = AccessToken.getInfoFromAccessTokenWithoutVerifying(token);
 
+        // TODO(mihaly) dont call getSession twice
         JsonObject newJWTUserPayload = userDataInJWT == null ? getSession(main, accessToken.sessionHandle).userDataInJWT
                 : userDataInJWT;
+        JsonObject newGrantPayload = grantPayload == null ? getSession(main, accessToken.sessionHandle).grants
+                : grantPayload;
+        if (newGrantPayload == null) {
+            newGrantPayload = new JsonObject();
+        }
```

Comment:
Some places, we handle the case that grapntPayload is null (in the plugin), but here you are normalising it to be an empty object. I think we should stick to one of those otherwise it can cause confusion.

Reply:
Yeah, because both regenerate and refresh "upgrades" the token to V3, which has a non-null `grantPayload`. I think the confusion comes from the fact that we continue to support V2 tokens (`grantPayload == null`) and V3 tokens.